Blog post on input checking (#150)
r-hub/blog: content/post/2021-10-07-input-checking/index.Rmd

---
slug: input-checking
title: "Checking the inputs of your R functions" 
authors: 

> - Sam Abbott

Reviewer: Y'all will need author files cf https://github.com/r-hub/blog/tree/master/content/authors

- Hugo Gruson
- Carl Pearson
- Tim Taylor
date: "2021-10-07" 
tags: 
- package development 
- r-package
output: hugodown::hugo_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(fig.path = "", comment = "")
# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)

```

## Introduction: the dangers of not checking function inputs

R functions and R packages are a convenient way to share code with the rest of the world
but it is generally not possible to know how, or with what precise aim in mind, others will
use your code. For example, they might try to use it on objects that your function was not
designed for. Let's imagine we have written a short function to compute the geometric mean:

```{r}

> geometric_mean <- function(...) {

Reviewer: this is a cool example!

## Checking function inputs using base R

There is a built-in mechanism to check input values in base R: `stopifnot()`.
You can see it [used](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/stats/R/approx.R#L78) [throughout](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/stats/R/cor.R#L36) [R](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/graphics/R/smoothScatter.R#L47) [source](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/srcfile.R#L23) [code](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/parse.R#L65).
As its name suggests, it will *stop* the function execution *if* an object does *not* pass some tests.

```{r, error = TRUE}
say_hello <- function(name) {
  stopifnot(is.character(name))
  paste("Hello", name)
}

say_hello("Bob")
say_hello(404)
```

However, as you can see in this example, the error message is not in plain English but contains some code instructions.

> This can hinder understanding of the issue.

Reviewer: Should https://style.tidyverse.org/error-messages.html be linked somewhere?

Author: It is linked just below, at line 109, as a footnote :wink:. Do you think it would be better here?

Reviewer: Then maybe more prominent in the text? :sweat_smile: :woman_facepalming:

Author: You mean I should promote the footnote to an actual sentence in the text?

Reviewer: yes, but you are free to disagree!

Author: I do think it fits better as a footnote. This is not exactly part of the main topic of the post but more like an aside.

This way, you don't have to repeat yourself which is generally a good aim [^2].
This becomes necessary when you start having many input checks in your function or in your package.

[^2]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)

## Checking function inputs using R packages

### The example of the checkmate package

Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:

```{r}
say_hello <- function(name) {
  # Among other things, check_string() checks that we provide a 
  # character object of length one

>   checkmate::assert_string(name)

Reviewer: can the message be tweaked by the way? Isn't "assertion" a potentially unknown word for some users?

Author: I don't think so. I think these packages have to be opinionated. If they leave too much flexibility to the user, we end up with `stopifnot()` or writing your own function.

  paste("Hello", name)
}
```

```{r, error = TRUE}
say_hello(404)
```

```{r, error = TRUE}
say_hello(c("Bob", "Alice"))
```

> ### Other packages to check function inputs

Reviewer: I was also wondering whether, for some packages that are not testthat, the number of revdeps could be interesting?

Author: Oh well, they are actually all on CRAN expected check. I might remove check (which is interesting but might have an overly complex syntax for a short demo) and add revdeps.

Author: I keep going back and forth with this but they all have 0 revdeps, excepted asserthat, which has 11 revdeps. I think I will not include this.

Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
We will not get into the details of all of the available options here but below is a list of some of the them. If interested in understanding the various approaches to input taking the documentation
for these package is a great place to start.

> -   testthat

Reviewer: link all of them? Maybe `{testthat}` in the Rmd will do the trick.

Author: Auto-linking with downlit (`{pkg}`) doesn't seem to work. I don't know why...

I manually added a link for each package.

> -   testthat

Reviewer: what in testthat by the way?

Reviewer 2: I agree - test that is out of place in this list, I'd suggest

Author: My understanding is that assertthat was archived precisely because all its function was integrated in testthat, which is more actively developed.

It's a somewhat uncommon use case but you case use testthat outside of the testting context. E.g.,

``` r
test <- function(x) {
  
  testthat::expect_length(x, 1)
  
}

test(c(1,2))
#> Error: `x` has length 2, not length 1.
```

<sup>Created on 2022-01-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Author: Your comments make me realise it may be nice to have a very small example for each package. So readers can get a feeling of how each package works. What do you think?

Reviewer: If not too much work, yes that might be good, or a _link_ to an example in the wild e.g. from github.com/cran? E.g. are there many packages using testthat in this way?

Author: I'll remove testthat. I do think it has its place here but I agree it can be confusing because it's not the primary focus of this package.

-   assertthat
-   check
-   assertr
-   assertive
-   ensurer
-   `vctrs::vec_assert()`

> 

Reviewer: In the post I'd be curious to also hear about your motivation to write this and a short paragraph or so about what you ended up using in your package(s) and whether you got some user feedback.

Author: In short: some of use were tired of filling our functions with `stopifnot()` / `if () { stop() }` or writing the same kind of checking function over and over and we started collectively looking for existing alternatives.

In which part of the post do you think this could fit?

Reviewer: Maybe right before the introduction? Something like

> Are you, like we were, tired of filling your functions with blabla?

## What about the future?

In this post, we have discussed some methods to check function inputs, and to generate more informative error messages when doing so.
However, this always comes with a performance cost, even though it's often relatively limited.
Zero-cost assertions, as found in some other languages, would require some kind of typing system which R does not currently support.

> Interestingly several other languages have evolved to have typing systems as they have developed (TypeScript as an extension of JavaScript, type annotations in Python).

Reviewer 2: Thanks all for this great post! One very minor comment: That final sentence is crucial to understanding broader implications of what you've set out, and would work better without the refs to js and python in parentheses, and perhaps slightly tweaking to make more active. Maybe something like:
> ... typing systems as they have developed. Typescript developed as an extension of JavaScript, and type annotations are now possible in python. 

I like the final link! :smirk:

Author: https://github.com/r-hub/blog/pull/150/commits/e5e9394c039f1b0b79c129b2799236a0250ea4cb

For example, they might try to use it on objects that your function was not designed for.
Let's imagine we have written a short function to compute the geometric mean:

```{r}
geometric_mean <- function(...) {
  
  return(prod(...)^(1/...length()))
  
}
```

When you tested the function yourself, everything seemed fine:

```{r}
geometric_mean(2, 8)

geometric_mean(4, 1, 1/32)
```

But a different person using your function might expose it to situations it was not prepared to handle, resulting in cryptic errors or undefined behaviour:

```{r, error = TRUE}
# Input with factors instead of numerics
geometric_mean(factor(2), 8)

# Input with negative values
geometric_mean(-1, 5)

# Input with NAs
geometric_mean(2, 8, NA)
```

Or worse, it could give an incorrect output:

```{r}
geometric_mean(c(2, 8))
```

Because of this, you need to make sure you return clear errors whenever your functions receives input it was not designed for.
In this blog post, we review a range of approaches to help you check your function inputs and discuss some potential future developments.

> ## Pre-requisite: thouroughly document your argument types

Reviewer: ```suggestion
## Pre-requisite: thoroughly document your argument types
```

You can notice from the simple example above that it's easy to pass invalid inputs to the `geometric_mean()` function because we didn't provide any documentation on what is or isn't a valid input.
We won't go into details here but the [roxygen2](https://roxygen2.r-lib.org/) package provides a convenient way to generate documentation for R functions.
Try to be as precise as possible when describing the required format for your inputs [^1].

[^1]: [Some package developers even developed their own standardized way to document argument types and length](https://github.com/r-lib/withr/commit/42e503092046705f30032cb3a321d64b0e9383d4).
    But there is currently no standard shared across the R community.

```{r}
#' @param name A character of length one with the name of the person to greet
say_hello <- function(name) {
  stopifnot(is.character(name))
  paste("Hello", name)
}
```

Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.

> ## Checking function inputs using base R

Reviewer: Maybe also mention the rlang "equivalents"? https://rlang.r-lib.org/reference/index.html#check-arguments

[^2]: Read [the tidyverse style guide](https://style.tidyverse.org/error-messages.html) for more guidance on how to write good error messages.

```{r, error = TRUE}
say_hello <- function(name) {
  stopifnot("`name` must be a character." = is.character(name))
  paste("Hello", name)
}

say_hello(404)
```

This is clearly a really great improvement to the functionality of base R.
However, we can see from this example that we could create the error message programmatically based on the contents of the test.
Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
This way, you don't have to repeat yourself which is generally a good aim [^3].
This becomes necessary when you start having many input checks in your function or in your package.

[^3]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)

## Checking function inputs using R packages

### The example of the checkmate package

> Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.

Reviewer: Link https://blog.r-hub.io/2019/12/12/internal-functions/? :-)

One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).

> checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:

Reviewer: ```suggestion
checkmate provides a large number of functions that check that inputs respect a given set of properties, and that return clear error messages when that is not the case:
```

```{r}
say_hello <- function(name) {
  # Among other things, check_string() checks that we provide a 
  # character object of length one
  checkmate::assert_string(name)
  paste("Hello", name)
}
```

```{r, error = TRUE}
say_hello(404)
```

```{r, error = TRUE}
say_hello(c("Bob", "Alice"))
```

### Other packages to check function inputs

Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
We will not get into the details of all of the available options here but below is a list of some of them, listed by decreasing number of reverse dependencies.
If interested in understanding the various approaches to input taking the documentation for these package is a great place to start.

-   [assertthat](https://github.com/hadley/assertthat)

```{r, error = TRUE}
assertthat::assert_that(is.character(1))
```

> -   [assertr](https://docs.ropensci.org/assertr/)

Reviewer: Interesting, I suppose it's an use case indeed, but I thought about it as an alternative to pointblank. :thinking:

```{r, error = TRUE}
library(magrittr)

mtcars %>%
  assertr::verify(nrow(.) < 10)
```

-   [assertive](https://bitbucket.org/richierocks/assertive)

```{r, error = TRUE}
assertive::assert_is_a_string(1)
```

-   [ensurer](https://github.com/smbache/ensurer)

```{r, error = TRUE}
ensure_square <- ensurer::ensures_that(NCOL(.) == NROW(.))

ensure_square(matrix(1:20, 4, 5))
```

-   `vctrs::vec_assert()`

```{r, error = TRUE}

> vctrs::vec_assert(c(1, 2), "character")

Reviewer: ça rend bizarre dans la preview (locale, `hugodown::hugo_start()`), y a-t-il une option à changer quand on l'utilise dans R Markdown ?

Reviewer: Maybe the `NO_COLOR` environment variable https://cli.r-lib.org/reference/faq.html

vctrs::vec_assert(c(1, 2), size = 3)
```

> -   [check](https://github.com/moodymudskipper/check) is slightly different because it doesn't provide utilities that work out of the box, but rather tools to assist you in writing your own checking functions

Reviewer: Is there any use case to mention here?

---
slug: input-checking
title: "Checking the inputs of your R functions" 
authors: 
- Hugo Gruson
- Sam Abbott
- Carl Pearson
date: "2021-10-07" 
tags: 
- package development 
- r-package
output: hugodown::hugo_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(fig.path = "", comment = "")
# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)
options(crayon.enabled = FALSE)

```

Are you, like we were, tired of filling your functions with argument checking code that sometimes ends up being longer that the core of the function itself?
Are you trying to find what is the most efficient approach to check inputs easily and without forgetting any edge cases?
Read about our exploration into the various ways to check your function inputs in R in this blog post.

> And please share your own tips and discoveries in the comment section below!

Reviewer: ```suggestion
And please share your own tips and discoveries in the comment section!
```

## Pre-requisite: thoroughly document your argument types

You can notice from the simple example above that it's easy to pass invalid inputs to the `geometric_mean()` function because we didn't provide any documentation on what is or isn't a valid input.
We won't go into details here but the [roxygen2](https://roxygen2.r-lib.org/) package provides a convenient way to generate documentation for R functions.
Try to be as precise as possible when describing the required format for your inputs [^1].

[^1]: [Some package developers even developed their own standardized way to document argument types and length](https://github.com/r-lib/withr/commit/42e503092046705f30032cb3a321d64b0e9383d4).
    But there is currently no standard shared across the R community.

```{r}
#' @param name A character of length one with the name of the person to greet
say_hello <- function(name) {
  stopifnot(is.character(name))
  paste("Hello", name)
}
```

Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.

## Checking function inputs using base R [^2]

> [^2]: Note that these base functions have equivalent in the tidyverse with a more consistent design and coloured output.

Reviewer: In any case if you keep it as a footnote it'd be good if the footnote came _after_ explaining what the base R functions are.

    `match.arg()`'s equivalent is `rlang::arg_match()` and `stopifnot()`'s

### `match.arg()`

If the input can only take specific values, the base function `match.arg()` can also prove useful:

```{r, error = TRUE}
match.arg(arg = "R", choices = c("R", "python"))

match.arg(arg = "javascript", choices = c("R", "python"))
```

But the real power of the `match.arg()` function comes from the fact that `choices` can be automatically obtained in the context of a function:

```{r, error = TRUE}

> choose_language <- function(language = c("R", "python")) {

Reviewer: One thing that's not obvious to me when using this pattern is what the default value is. Could you add a sentence about this?

Reviewer: (I know it _should_ be obvious to me :sweat_smile: :see_no_evil: )

This can hinder understanding of the issue.

Because of this, `stopifnot()` was improved in R 4.0.0:

> stopifnot() now allows customizing error messages via argument names, thanks to a patch proposal by Neal Fultz in PR#17688.

This means we can now provide a clearer error message directly in `stopifnot()` [^3]:

[^3]: Read [the tidyverse style guide](https://style.tidyverse.org/error-messages.html) for more guidance on how to write good error messages.

```{r, error = TRUE}
say_hello <- function(name) {
  stopifnot("`name` must be a character." = is.character(name))
  paste("Hello", name)
}

say_hello(404)
```

This is clearly a really great improvement to the functionality of base R.
However, we can see from this example that we could create the error message programmatically based on the contents of the test.
Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
This way, you don't have to repeat yourself which is generally a good aim [^4].
This becomes necessary when you start having many input checks in your function or in your package.

> [^4]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)

Reviewer: ```suggestion
[^4]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](/2021/07/30/cache/)
```

## Checking function inputs using R packages

### The example of the checkmate package

Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem [^5], you can also rely on existing packages to make your life easier.
One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
checkmate provides a large number of functions that check that inputs respect a given set of properties, and that return clear error messages when that is not the case:

> [^5]: See [this earlier blog post](https://blog.r-hub.io/2019/12/12/internal-functions/) for more information about why and who you would go with writing internal functions.

Reviewer: ```suggestion
[^5]: See [this earlier blog post](/2019/12/12/internal-functions/) for more information about why and who you would go with writing internal functions.
```

Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
We will not get into the details of all of the available options here but below is a list of some of them, listed by decreasing number of reverse dependencies.
If interested in understanding the various approaches to input taking the documentation for these package is a great place to start.

-   [assertthat](https://github.com/hadley/assertthat)

```{r, error = TRUE}
assertthat::assert_that(is.character(1))
```

-   [assertr](https://docs.ropensci.org/assertr/)

```{r, error = TRUE}
library(magrittr)

mtcars %>%
  assertr::verify(nrow(.) < 10)
```

-   [assertive](https://bitbucket.org/richierocks/assertive)

```{r, error = TRUE}
assertive::assert_is_a_string(1)
```

-   [ensurer](https://github.com/smbache/ensurer)

```{r, error = TRUE}
ensure_square <- ensurer::ensures_that(NCOL(.) == NROW(.))

ensure_square(matrix(1:20, 4, 5))
```

-   `vctrs::vec_assert()`

```{r, error = TRUE}
vctrs::vec_assert(c(1, 2), "character")

vctrs::vec_assert(c(1, 2), size = 3)
```

> -   [check](https://github.com/moodymudskipper/check) is slightly different because it doesn't provide utilities that work out of the box, but rather tools to assist you in writing your own checking functions

Reviewer: Could you also add a link to a package using it?

Author: I can't find any :/

In this post, we have discussed some methods to check function inputs, and to generate more informative error messages when doing so.
However, this always comes with a performance cost, even though it's often relatively limited.
Zero-cost assertions, as found in some other languages, would require some kind of typing system which R does not currently support.

> Interestingly several other languages have evolved to havetyping systems as they have developed.

Reviewer: ```suggestion
Interestingly several other languages have evolved to have typing systems as they have developed.
```
